en/ros/ros2_offboard_control.md

```diff
@@ -1,172 +1,162 @@
+After that we can follow a similar set of steps to those in [ROS 2 User Guide > Build ROS2 Workspace](..ros/ros2_comm.md#build-ros-2-workspace) to run the example.
+
+To build and run the example:
+
+1. Open a new terminal.
+1. Create and navigate into a new colcon workspace directory using:
+
+   ```sh
+   mkdir -p ~/ws_offboard_control/src/
+   cd ~/ws_offboard_control/src/
    ```
 
-   In *path_to_colcon_ws/src/px4_ros_com/templates/urtps_bridge_topics.yaml*:
-   ```yaml
-   - msg:     OffboardControlMode
-     receive: true
-   ...
-   - msg:     VehicleCommmand
-     receive: true
-   ...
-   - msg:     VehicleLocalPositionSetpoint
-     receive: true
-   - msg:     TrajectorySetpoint
-     base:    VehicleLocalPositionSetpoint
-     receive: true
+1. Clone the [px4_msgs](https://github.com/PX4/px4_msgs) repo to the `/src` directory (this repo is needed in every ROS2 PX4 workspace!):
+
+   ```sh
+   git clone https://github.com/PX4/px4_msgs.git
+   # checkout the matching release branch if not using PX4 main.
    ```
- 1. `vehicle_command` message is configured in the `urtps_bridge_topics.yaml` file both in the PX4-Autopilot and
- *px4_ros_com* package to *send* to the Autopilot.
 
-    In *PX4-Autopilot/msg/tools/urtps_bridge_topics.yaml*:
-    ```yaml
-    - msg:     vehicle_command
-      receive: true
-    ```
+1. Clone the example repository [px4_ros_com](https://github.com/PX4/px4_ros_com) to the `/src` directory:
 
-    In *path_to_colcon_ws/src/px4_ros_com/templates/uorb_rtps_message_ids.yaml*:
-    ```yaml
-    - msg:     VehicleCommmand
-      receive: true
-    ```
+   ```sh
+   git clone https://github.com/PX4/px4_ros_com.git
+   ```
 
-   :::note
-   At time of writing, the above topic is already configured to be sent.
-   :::
+1. Source the ROS2 development environment ("foxy") into the current terminal and compile the workspace using `colcon`:
 
-## Implementation
+   ```sh
+   cd ..
+   source /opt/ros/foxy/setup.bash
+   colcon build
+   ```
 
-The source code of the offboard control example can be found in [offboard_control.cpp](https://github.com/PX4/px4_ros_com/blob/main/src/examples/offboard/offboard_control.cpp).
+1. Source the `local_setup.bash`:
 
-Here are some details about the implementation:
+   ```sh
+   source install/local_setup.bash
+   ```
+1. Launch the example.
 
-```cpp
-timesync_sub_ = this->create_subscription<px4_msgs::msg::Timesync>("/fmu/timesync/out",
-    10,
-    [this](const px4_msgs::msg::Timesync::UniquePtr msg) {
-        timestamp_.store(msg->timestamp);
-    });
-```
+   ```
+   ros2 run px4_ros_com offboard_control
+   ```
 
-The above is required in order to obtain a synchronized timestamp to be set and sent with the `offboard_control_mode` and `trajectory_setpoint` messages.
+The vehicle should arm, ascend 5 metres, and then wait (perpetually).
+
+## Implementation
```

Comment:
FYI the main change here is to explain a little more about what offboard mode requires, and link to that. The idea being that first you understand why this code writes setpoints before changing the mode, and also that you can go to offboard mode to understand how you might do velocity control etc.

Reply:
I went through the code and instructions, everything works just fine but
and you won't like it
1. It seems that you can switch to offboard mode anytime.
2. Arming is disabled until a valid stream of setpoints is received (as it should be).

To be more precise, what I did was

1. start the simulation and open QGC.
2. Without even starting the micro XRCE-DDS Agent I switched to offboard mode from QGC. The command was accepted. I also tried from mavlink console with `commander mode offboard` and the result was the same.
3. With mode set to offboard arming was denied. The reported description in QGC is _The offboard component is not sending setpoints or the required estimate (e.g. position) is missing._. This actually tells us that contrarily to what is written in the first note of [offboard-mode](https://docs.px4.io/main/en/flight_modes/offboard.html#offboard-mode), only an estimate of what we want to control is needed. For example, If I want to send attitude setpoints I can do it without GPS.
4. As soon as I started sending setpoints arming was allowed again.

Further testing made me realize also that you only need to provide a continuous stream of `offboard_control_mode` messages to allow arming. Therefore you can do weird things like

1. Start streaming `offboard_control_mode` messages (for example in offboard position mode)
2. Arm the vehicle.
3. After a while (before the auto disarm is triggered) start sending trajectory setpoints. The vehicle will take off.
4. Stop sending trajectory setpoints, the vehicle will use the last one available and keep flying. No error is raised.

Reply:
PX4 should never allow you to switch to offboard mode without sending a stream of offboard_control_mode - that's like the heartbeat. If offboard mode messages stop it should switch to hold mode.

So your further testing is correct as per my understanding. I think using the last trajectory setpoint is OK for position setpoints because you'll fly to that position and then stop. If the same thing happens for velocity setpoints it is risky -though as above, I think ONLY the offboard control mode  is the heartbeat. It may be a "problem", but it is expected behaviour.

Re 
> he offboard component is not sending setpoints or the required estimate (e.g. position) is missing.. 

Hmmm. You're likely right. If you're in offboard mode the assumption is that position is provided by the companion - PX4 has no need of its own position/GPS.

@Jaeyoung-Lim Does this align with your understanding too?

Reply:
@beniaminopozzan FYI a little more debug (some duplicated) which pretty much confirm my assumptions
1. `OffboardControlMode` must be streamed to arm in offboard mode _OR to switch to offboard mode when flying_. If this stops, you'll drop back to position mode by default (the fallback mode is configurable by parameters). What this means is that this is the heartbeat, and does block offboard operation if not sent (though you can enter the mode).
2. The setpoints make no difference to mode switching.

IN theory if you have RC required you should also need https://docs.px4.io/main/en/advanced_config/parameter_reference.html#COM_RCL_EXCEPT set to avoid a failsafe. Julian had this in his tutorial. My testing is that this does not seem to matter - it may be that RC checks are disabled on simulator for some reason.

I'll update this doc with minimal changes, and do some updates to offboard mode as a separate PR.

Reply:
@hamishwillee Thanks for confirming it!
Did you have QGC with virtual joystick opened? I would explain why it had no effects.

Reply:
No. I thought the same thing so I tried with COM_RC_IN_MODE = 0 (joystick only), and also checked that there was no virtual joystick. This was a completely clean installation on a fresh VM. I suspect the check is broken.